epochStart/metachain/systemSCs.go

```diff
@@ -327,9 +333,134 @@ func (s *systemSCProcessor) ProcessSystemSmartContract(
 		}
 	}
 
+	if s.flagStakingV4Enabled.IsSet() {
+		allNodesKeys := s.getAllNodeKeys(validatorInfos)
+
+		err := s.stakingDataProvider.PrepareStakingData(allNodesKeys)
+		if err != nil {
+			return err
+		}
+
+		err = s.selectNodesFromAuctionList(validatorInfos, randomness)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *systemSCProcessor) selectNodesFromAuctionList(validatorInfos map[uint32][]*state.ValidatorInfo, randomness []byte) error {
+	auctionList, noOfValidators := getAuctionListAndNoOfValidators(validatorInfos)
+	err := s.sortAuctionList(auctionList, randomness)
+	if err != nil {
+		return err
+	}
+
+	auctionListSize := uint32(len(auctionList))
+	noOfAvailableNodeSlots := core.MinUint32(auctionListSize, s.maxNodes-noOfValidators)
+	s.displayAuctionList(auctionList, noOfAvailableNodeSlots)
+
+	for i := uint32(0); i < noOfAvailableNodeSlots; i++ {
+		auctionList[i].List = string(common.NewList)
+	}
+
+	return nil
+}
+
+func getAuctionListAndNoOfValidators(validatorInfos map[uint32][]*state.ValidatorInfo) ([]*state.ValidatorInfo, uint32) {
```

Comment:
info does not have a plural :D
validatorInfoMap ?

Reply:
Right, I've copy-pasted the param name from other funcs in this file. Apparently, all of them use **validatorInfos**